indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentAllocateAction.java

```diff
@@ -0,0 +1,268 @@
+
+    final List<Interval> tryIntervals = Lists.newArrayList();
+
+    final Interval rowInterval = new Interval(
+        queryGranularity.truncate(timestamp.getMillis()),
+        queryGranularity.next(queryGranularity.truncate(timestamp.getMillis()))
+    );
+
+    final Set<DataSegment> usedSegmentsForRow = ImmutableSet.copyOf(
+        msc.getUsedSegmentsForInterval(dataSource, rowInterval)
+    );
+
+    if (usedSegmentsForRow.isEmpty()) {
+      // No existing segments for this row, but there might still be nearby ones that conflict with our preferred
+      // segment granularity. Try that first, and then progressively smaller ones if it fails.
+      for (Granularity gran : granularitiesFinerThan(preferredSegmentGranularity)) {
+        tryIntervals.add(gran.bucket(timestamp));
+      }
+    } else {
+      // Existing segment(s) exist for this row; use the interval of the first one.
+      tryIntervals.add(usedSegmentsForRow.iterator().next().getInterval());
+    }
+
+    for (final Interval tryInterval : tryIntervals) {
+      if (tryInterval.contains(rowInterval)) {
+        log.debug(
+            "Trying to allocate pending segment for rowInterval[%s], segmentInterval[%s].",
+            rowInterval,
+            tryInterval
+        );
+        final TaskLock tryLock = toolbox.getTaskLockbox().tryLock(task, tryInterval).orNull();
+        if (tryLock != null) {
+          final SegmentIdentifier identifier = msc.allocatePendingSegment(
+              dataSource,
+              sequenceName,
+              previousSegmentId,
+              tryInterval,
+              tryLock.getVersion()
+          );
+          if (identifier != null) {
+            return Optional.of(identifier);
+          } else {
+            log.debug(
+                "Could not allocate pending segment for rowInterval[%s], segmentInterval[%s].",
+                rowInterval,
+                tryInterval
+            );
+          }
+        } else {
+          log.debug("Could not acquire lock for rowInterval[%s], segmentInterval[%s].", rowInterval, tryInterval);
+        }
+      }
+    }
+
+    // Could not allocate a pending segment. There's a chance that this is because someone else inserted a segment
+    // overlapping with this row between when we called "mdc.getUsedSegmentsForInterval" and now. Check it again,
+    // and if it's different, repeat.
+
+    if (!ImmutableSet.copyOf(msc.getUsedSegmentsForInterval(dataSource, rowInterval)).equals(usedSegmentsForRow)) {
+      final long shortRandomSleep = 50 + (long) (Math.random() * 450);
+      log.debug(
+          "Used segment set changed for rowInterval[%s]. Retrying segment allocation in %,dms.",
+          rowInterval,
+          shortRandomSleep
+      );
+      try {
+        Thread.sleep(shortRandomSleep);
+      }
+      catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw Throwables.propagate(e);
+      }
+      return perform(task, toolbox);
```

Comment:
nit: this can in theory become an infinite retry, can we do an upper bounded number retries by probably using while/for loop instead of recursion?